### PR TITLE
FPGA: Add sample for `max_reinvocation_delay` attribute

### PR DESCRIPTION
**Pull request by williamwang256**
# Adding a New Sample(s)

## Description

New sample demonstrates how and when to use the `max_reinvocation_delay` loop attribute.

## Checklist
Administrative
- [ ] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [ ] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [ ] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [ ] Adhere to readme template 
- [ ] Enforce format via clang-format config file
- [ ] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [ ] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [ ] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

**Review by yuguen-intel — APPROVED** [review]

**Comment by yuguen-intel** [comment]
@tomlenth and @pmpeter1 would be appreciated if you find the time to review this new FPGA code sample

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+# Use cmake -DFPGA_DEVICE=<board-support-package>:<board-variant> to choose a
+# different device. Here are a few device examples (this list is not
+# exhaustive):
+#   intel_s10sx_pac:pac_s10
+#   intel_s10sx_pac:pac_s10_usm
+#   intel_a10gx_pac:pac_a10
+# Note that depending on your installation, you may need to specify the full 
+# path to the board support package (BSP), this usually is in your install 
+# folder.
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/CMakeLists.txt:33]:
```suggestion
# Use cmake -DFPGA_DEVICE=<board-support-package>:<board-variant> to choose a
# different device. 
# Note that depending on your installation, you may need to specify the full 
# path to the board support package (BSP), this usually is in your install 
# folder.
```

**Review by yuguen-intel — CHANGES_REQUESTED** [review]
Removing mentions to the now unsupported PAC boards

**Review by yuguen-intel — APPROVED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
+#include <iostream>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+#define FACTORS 5
```
Comment (pmpeter1) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:7]: FACTORS should follow convention used in other samples of constexpr (not #define).

**Review by pmpeter1 — CHANGES_REQUESTED** [review]

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp (hunk 2 above)
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:7]:
```suggestion
constexpr int kFactors 5
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+// Forward declare the kernel and pipe names
+// (This prevents unwanted name mangling in the optimization report.)
+class ArithmeticSequence;
+class ResultsPipe;
+
+// Results pipe from device back to host
+using PipeResults = sycl::ext::intel::experimental::pipe<ResultsPipe, int>;
+
+// Computes and outputs the first "sequence_length" terms of the arithmetic
+// sequences with first term "first_term" and factors 1 through FACTORS.
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:18]:
```suggestion
// sequences with first term "first_term" and factors 1 through kFactors.
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+struct ArithmeticSequenceKernel {
+  int first_term;
+  int sequence_length;
+
+  void operator()() const {
+    for (int factor = 1; factor <= FACTORS; factor++) {
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:24]:
```suggestion
    for (int factor = 1; factor <= kFactors; factor++) {
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+    int first_term = 0;
+    int sequence_length = 10;
+
+    q.single_task<ArithmeticSequence>(
+        ArithmeticSequenceKernel{first_term, sequence_length});
+
+    // Verify functional correctness
+    bool passed = true;
+    for (int factor = 1; factor <= FACTORS; factor++) {
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:58]:
```suggestion
    for (int factor = 1; factor <= kFactors; factor++) {
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/README.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+```
+
+By default, the compiler inserts a loop reinvocation delay of a few cycles on the inner loop. In general, the compiler may insert a higher loop reinvocation delay to better optimize a loop by enabling loop speculation or by pipelining the loop orchestration hardware, which can increase fMAX. This delay can be seen by observing the transactions coming out of the streaming interface associated with `PipeOut`.
+
+![default behavior with delay between invocations](assets/default.png)
+
+However, the extra latency between invocations of a loop can have a significant impact in performance if the loop has a very small trip count. Applying `[[intel::max_reinvocation_delay(1)]]` to the inner loop allows us to remove this delay between invocations. 
+
+```c++
+for (int factor = 0; factor < FACTORS; factor++) {
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/README.md:82]:
```suggestion
for (int factor = 0; factor < kFactors; factor++) {
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/README.md [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+> **Note**:  A loop **invocation** begins when the program flow enters a loop for the first time, while an **iteration** begins each time the program flow enters the loop body. In the example below, the `i` loop will have `sequence_length` iterations for each invocation. It will be invoked once for each **iteration** of the outer `factor` loop.
+
+The attribute parameter `N` is required and must be a positive constant expression of integer type. This parameter controls the maximum loop reinvocation delay allowed, measured in clock cycles. Currently, only `N=1` is supported, i.e., that there should be no delay between invocations. If you do not specify this attribute, the compiler may insert some delay to improve fMAX as shown in the following example.
+
+### Example
+
+Consider the following loop nest:
+
+```c++
+for (int factor = 0; factor < FACTORS; factor++) {
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/README.md:68]:
```suggestion
for (int factor = 0; factor < kFactors; factor++) {
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp (hunk 2 above)
Comment (yuguen-intel) [reply DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp]: Now addressed.

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
+#include <iostream>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+constexpr int kFactors 5
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:7]:
```suggestion
constexpr int kFactors 5;
```

File: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp [hunk 9]
```diff
@@ -0,0 +1,86 @@
+#include <iostream>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+constexpr int kFactors 5;
```
Comment (yuguen-intel) [inline DirectProgramming/C++SYCL_FPGA/Tutorials/Features/max_reinvocation_delay/src/max_reinvocation_delay.cpp:7]:
```suggestion
constexpr int kFactors = 5;
```

**Review by ishaghosh27 — APPROVED** [review]

**Pull request merged**